vllm/entrypoints/openai/reasoning_parsers/deepseek_r1_reasoning_parser.py

```diff
@@ -124,25 +120,42 @@ def extract_reasoning_content_streaming(
     def extract_reasoning_content(
             self, model_output: str, request: ChatCompletionRequest
     ) -> tuple[Optional[str], Optional[str]]:
+        """
+        Extract reasoning content from the model output.
+
+        For text <think>abc</think>xyz:
+        - 'abc' goes to reasoning_content
+        - 'xyz' goes to content
+
+        Returns:
+            tuple[Optional[str], Optional[str]]: reasoning content and content
+        """
+
+        # Check if the start token is present in the model output, remove it
+        # if it is present.
+        start_token_index = model_output.find(self.think_start_token)
+        if start_token_index != -1:
+            model_output = model_output[start_token_index +
+                                        len(self.think_start_token):]
 
         # DeepSeek R1 doesn't generate <think> now.
         # Thus we assume the reasoning content is always at the start.
         # Ref https://huggingface.co/deepseek-ai/DeepSeek-R1/commit/8a58a132790c9935686eb97f042afa8013451c9f
         if self.think_end_token not in model_output:
             return model_output, None
         else:
-            # Add a start token if it's missing to keep compatibility.
-            if self.think_start_token not in model_output:
-                model_output = f"{self.think_start_token}{model_output}"
-            # Use a regex to find the reasoning content
-            reasoning_content = self.reasoning_regex.findall(model_output)[0]
-
-            end_index = len(
-                f"{self.think_start_token}{reasoning_content}{self.think_end_token}"
-            )
-            final_output = model_output[end_index:]
-
-            if len(final_output) == 0:
+            # Find the end token index in the model output.
+            end_token_index = model_output.find(self.think_end_token)
+            # If the end token is not found, return the model output as is.
+            # It should not happen since we already checked for the presence
+            # of the end token.
+            if end_token_index == -1:
+                return model_output, None
+            # Extract the reasoning content before the end token.
+            reasoning_content = model_output[:end_token_index]
+            # Extract the content after the end token.
+            content = model_output[end_token_index +
+                                   len(self.think_end_token):]
+            if len(content) == 0:
                 return reasoning_content, None
```

Comment:
Ditto, now the logic is simpler than the above as reasoning content is always at the first half.
```suggestion
            reasoning_content, think_end_token, content = model_output.partition(self.think_end_token)
            # If the end token is not found, return the model output as is.
            # It should not happen since we already checked for the presence
            # of the end token.
            # If generation stops right after end-of-think, return null content
            content = content or None
```

Reply:
I was thinking can't we just assume for the thinking models to always output reasoning first until we found `</think>`?

Reply:
@aarnphm  To maintain compatibility with the old chat template, we must locate the `<think>` token and extract the text between it and the corresponding closing token.

For the new chat template, we can assume `the thinking models to always output reasoning first until we found </think>`

Reply:
this is for deepseek r1, all chat templates will have `<think>`

Reply:
The old chat template does not have `<think>`. Thus the model will generate one. New chat template includes it, thus the model will generate reasoning content directly.

Reply:
Ref https://github.com/vllm-project/vllm/pull/13025

Reply:
We will always fetch the results from main from `snapshot_download`, so im certain that even if you are using r1 models, it should fetch the latest changes for the tokenizer.

Reply:
I think so at that time, but some users used `TRANSFORMERS_OFFLINE=1`, or just use local model from s3 with the old template.